### ISSUE TITLE
Panic on sysctl -a (v2.2.3rc3)

### ISSUE THREAD

**Issue by Haravikk** [openzfsonosx/openzfs#126]
### System information
Type | Version/Name
 --- | ---
Distribution Name	| macOS
Distribution Version	| 10.13.6
Linux Kernel	| darwin 17.7.0
Architecture	| x86_64
ZFS Version	| zfs-macOS-2.2.3-rc3

### Describe the problem you're observing

While testing v2.2.3rc3 on both a 2009 Mac Mini and a 2018 Mac Mini I encountered a kernel panic when running `sysctl -a`, seems like one of the keys when accessed is causing a panic somehow.

### Describe how to reproduce the problem

1. Install v2.2.3rc3 on an Intel Mac (2009 and 2018 Mac Minis tested)
2. Run `sysctl -a`

### Panic Log

I wasn't able to get a panic log from the 2018 Mac Mini (Catalina) for some reason, but the 2009 Mac Mini (High Sierra) gave me the following:

```
Anonymous UUID:       FF4F41DB-5BFD-2224-AD15-21436B434DB9

Mon Mar 18 20:27:12 2024

*** Panic Report ***
panic(cpu 1 caller 0xffffff800e38687f): Kernel trap at 0xffffff800e27d800, type 14=page fault, registers:
CR0: 0x0000000080010033, CR2: 0x0000000092313517, CR3: 0x000000008a7ce000, CR4: 0x00000000000026e0
RAX: 0xebc6fe2778a50085, RBX: 0x0000000000000000, RCX: 0xffffff90ac4c38c0, RDX: 0xffffff800e743610
RSP: 0xffffff90ac4c37e0, RBP: 0xffffff90ac4c38b0, RSI: 0x0000000092313517, RDI: 0x0000000092313517
R8:  0x000000000000000a, R9:  0x0000000000000000, R10: 0xffffff800e743610, R11: 0x0000000000000000
R12: 0xffffff90ac4c3910, R13: 0x00000000ffffffff, R14: 0xffffff90ac4c38c0, R15: 0xffffff800e27e574
RFL: 0x0000000000010246, RIP: 0xffffff800e27d800, CS:  0x0000000000000008, SS:  0x0000000000000010
Fault CR2: 0x0000000092313517, Error code: 0x0000000000000000, Fault CPU: 0x1, PL: 0, VF: 1

Backtrace (CPU 1), Frame : Return Address
0xffffff90ac4c32b0 : 0xffffff800e26ae76 
0xffffff90ac4c3300 : 0xffffff800e394484 
0xffffff90ac4c3340 : 0xffffff800e386654 
0xffffff90ac4c33b0 : 0xffffff800e21ce60 
0xffffff90ac4c33d0 : 0xffffff800e26a8ec 
0xffffff90ac4c3500 : 0xffffff800e26a6ac 
0xffffff90ac4c3560 : 0xffffff800e38687f 
0xffffff90ac4c36d0 : 0xffffff800e21ce60 
0xffffff90ac4c36f0 : 0xffffff800e27d800 
0xffffff90ac4c38b0 : 0xffffff800e743705 
0xffffff90ac4c38d0 : 0xffffff7f9204d995 
0xffffff90ac4c3940 : 0xffffff7f921f5f54 
0xffffff90ac4c3d90 : 0xffffff800e731676 
0xffffff90ac4c3e10 : 0xffffff800e7318a1 
0xffffff90ac4c3f40 : 0xffffff800e802818 
0xffffff90ac4c3fa0 : 0xffffff800e21d646 
      Kernel Extensions in backtrace:
         org.openzfsonosx.zfs(2.2.3)[6FD41A71-ADD3-362E-98A2-9D55FC6651BC]@0xffffff7f92045000->0xffffff7f92771fff
            dependency: com.apple.iokit.IOStorageFamily(2.1)[F27A8A2A-6662-3608-83BD-415037509E01]@0xffffff7f8ec6f000

BSD process name corresponding to current thread: sysctl

Mac OS version:
17G14042

Kernel version:
Darwin Kernel Version 17.7.0: Fri Oct 30 13:34:27 PDT 2020; root:xnu-4570.71.82.8~1/RELEASE_X86_64
Kernel UUID: 5E83A13A-32F5-3604-8591-50E2F2F70DC6
Kernel slide:     0x000000000e000000
Kernel text base: 0xffffff800e200000
__HIB  text base: 0xffffff800e100000
System model name: Macmini4,1 (Mac-F2208EC8)

System uptime in nanoseconds: 199262564347318
last loaded kext at 199225121448855: com.apple.driver.AppleXsanScheme	3 (addr 0xffffff7f92772000, size 32768)
last unloaded kext at 199029452309341: org.openzfsonosx.zfs	2.2.3 (addr 0xffffff7f8ec9f000, size 7520256)
loaded kexts:
org.openzfsonosx.zfs	2.2.3
com.apple.driver.AppleBluetoothMultitouch	92
com.apple.driver.AudioAUUC	1.70
com.apple.driver.AppleHWSensor	1.9.5d0
com.apple.driver.AGPM	110.23.37
com.apple.driver.ApplePlatformEnabler	2.7.0d0
com.apple.filesystems.autofs	3.0
com.apple.driver.AppleMikeyHIDDriver	131
com.apple.driver.AppleHDA	281.52
com.apple.driver.AppleUpstreamUserClient	3.6.5
com.apple.driver.AppleMCCSControl	1.5.5
com.apple.AGDCPluginDisplayMetrics	3.20.18
com.apple.GeForceTesla	10.0.4
com.apple.driver.AppleHV	1
com.apple.iokit.IOUserEthernet	1.0.1
com.apple.iokit.IOBluetoothSerialManager	6.0.7f22
com.apple.driver.AppleMikeyDriver	281.52
com.apple.driver.pmtelemetry	1
com.apple.Dont_Steal_Mac_OS_X	7.0.0
com.apple.driver.AppleIntelSlowAdaptiveClocking	4.0.0
com.apple.driver.AppleOSXWatchdog	1
com.apple.driver.AppleLPC	3.1
com.apple.driver.ACPI_SMC_PlatformPlugin	1.0.0
com.apple.driver.AppleIRController	353
com.apple.filesystems.hfs.kext	407.50.6
com.apple.AppleFSCompression.AppleFSCompressionTypeDataless	1.0.0d1
com.apple.BootCache	40
com.apple.AppleFSCompression.AppleFSCompressionTypeZlib	1.0.0
com.apple.AppleSystemPolicy	1.0
com.apple.iokit.SCSITaskUserClient	404.30.3
com.apple.filesystems.apfs	748.51.0
com.apple.driver.AppleFWOHCI	5.5.9
com.apple.driver.AirPort.Brcm4331	800.21.30
com.apple.driver.AppleSDXC	1.7.6
com.apple.iokit.AppleBCM5701Ethernet	10.3.1
com.apple.driver.AppleAHCIPort	329.50.2
com.apple.driver.AppleRTC	2.0
com.apple.driver.AppleHPET	1.8
com.apple.driver.AppleACPIButtons	6.1
com.apple.driver.AppleSMBIOS	2.1
com.apple.driver.AppleACPIEC	6.1
com.apple.driver.AppleAPIC	1.7
com.apple.driver.AppleIntelCPUPowerManagementClient	220.50.1
com.apple.nke.applicationfirewall	186
com.apple.security.TMSafetyNet	8
com.apple.security.quarantine	3
com.apple.driver.AppleIntelCPUPowerManagement	220.50.1
com.apple.driver.AppleXsanScheme	3
com.apple.driver.IOBluetoothHIDDriver	6.0.7f22
com.apple.driver.AppleMultitouchDriver	1614.1
com.apple.driver.AppleInputDeviceSupport	1614.13
com.apple.kext.triggers	1.0
com.apple.driver.DspFuncLib	281.52
com.apple.kext.OSvKernDSPLib	526
com.apple.nvidia.classic.NVDANV50HalTesla	10.0.4
com.apple.nvidia.classic.NVDAResmanTesla	10.0.4
com.apple.iokit.IOAVBFamily	683.1
com.apple.plugin.IOgPTPPlugin	680.15
com.apple.driver.AppleSSE	1.0
com.apple.iokit.IOSurface	211.15
com.apple.iokit.IOSerialFamily	11
com.apple.AppleGPUWrangler	3.20.18
com.apple.AppleGraphicsDeviceControl	3.20.18
com.apple.iokit.IOSlowAdaptiveClockingFamily	1.0.0
com.apple.iokit.IOFireWireIP	2.2.9
com.apple.driver.IOPlatformPluginLegacy	1.0.0
com.apple.driver.IOPlatformPluginFamily	6.0.0d8
com.apple.driver.AppleHDAController	281.52
com.apple.iokit.IOHDAFamily	281.52
com.apple.iokit.IOAudioFamily	206.5
com.apple.vecLib.kext	1.2.0
com.apple.driver.AppleSMBusController	1.0.18d1
com.apple.driver.AppleSMBusPCI	1.0.14d1
com.apple.iokit.IONDRVSupport	519.21
com.apple.iokit.IOGraphicsFamily	519.23
com.apple.iokit.BroadcomBluetoothHostControllerUSBTransport	6.0.7f22
com.apple.iokit.IOBluetoothHostControllerUSBTransport	6.0.7f22
com.apple.iokit.IOBluetoothHostControllerTransport	6.0.7f22
com.apple.iokit.IOBluetoothFamily	6.0.7f22
com.apple.driver.AppleHIDKeyboard	205.1
com.apple.driver.usb.IOUSBHostHIDDevice	1.2
com.apple.driver.usb.AppleUSBHub	1.2
com.apple.driver.usb.networking	5.0.0
com.apple.driver.usb.AppleUSBHostCompositeDevice	1.2
com.apple.driver.CoreStorage	546.50.1
com.apple.filesystems.hfs.encodings.kext	1
com.apple.iokit.IOSCSIMultimediaCommandsDevice	404.30.3
com.apple.iokit.IOBDStorageFamily	1.8
com.apple.iokit.IODVDStorageFamily	1.8
com.apple.iokit.IOCDStorageFamily	1.8
com.apple.iokit.IOAHCISerialATAPI	267.50.1
com.apple.iokit.IOAHCIBlockStorage	301.40.2
com.apple.iokit.IOFireWireFamily	4.7.2
com.apple.iokit.IO80211Family	1200.12.2
com.apple.driver.corecapture	1.0.4
com.apple.iokit.IOEthernetAVBController	1.1.0
com.apple.driver.mDNSOffloadUserClient	1.0.1b8
com.apple.iokit.IOAHCIFamily	288
com.apple.driver.usb.AppleUSBEHCIPCI	1.2
com.apple.driver.usb.AppleUSBOHCIPCI	1.2
com.apple.driver.usb.AppleUSBOHCI	1.2
com.apple.driver.usb.AppleUSBEHCI	1.2
com.apple.driver.NVSMU	2.2.9
com.apple.driver.usb.AppleUSBHostPacketFilter	1.0
com.apple.iokit.IOUSBFamily	900.4.1
com.apple.driver.AppleUSBHostMergeProperties	1.2
com.apple.driver.AppleEFINVRAM	2.1
com.apple.driver.AppleEFIRuntime	2.1
com.apple.iokit.IOSMBusFamily	1.1
com.apple.iokit.IOHIDFamily	2.0.0
com.apple.security.sandbox	300.0
com.apple.kext.AppleMatch	1.0.0d1
com.apple.driver.DiskImages	480.60.3
com.apple.driver.AppleFDEKeyStore	28.30
com.apple.driver.AppleEffaceableStorage	1.0
com.apple.driver.AppleKeyStore	2
com.apple.driver.AppleUSBTDM	439.70.3
com.apple.driver.AppleMobileFileIntegrity	1.0.5
com.apple.iokit.IOUSBMassStorageDriver	140.70.2
com.apple.iokit.IOSCSIBlockCommandsDevice	404.30.3
com.apple.iokit.IOSCSIArchitectureModelFamily	404.30.3
com.apple.iokit.IOStorageFamily	2.1
com.apple.driver.AppleCredentialManager	1.0
com.apple.driver.KernelRelayHost	1
com.apple.iokit.IOUSBHostFamily	1.2
com.apple.driver.usb.AppleUSBCommon	1.0
com.apple.driver.AppleBusPowerController	1.0
com.apple.driver.AppleSEPManager	1.0.1
com.apple.driver.IOSlaveProcessor	1
com.apple.iokit.IOReportFamily	31
com.apple.iokit.IOTimeSyncFamily	680.15
com.apple.iokit.IONetworkingFamily	3.4
com.apple.driver.AppleACPIPlatform	6.1
com.apple.driver.AppleSMC	3.1.9
com.apple.iokit.IOPCIFamily	2.9
com.apple.iokit.IOACPIFamily	1.4
com.apple.kec.pthread	1
com.apple.kec.Libm	1
com.apple.kec.corecrypto	1.0

EOF
Model: Macmini4,1, BootROM 76.0.0.0.0, 2 processors, Intel Core 2 Duo, 2.4 GHz, 4 GB, SMC 1.65f2
Graphics: NVIDIA GeForce 320M, NVIDIA GeForce 320M
Memory Module: BANK 0/DIMM0, 2 GB, DDR3, 1067 MHz, 0x80CE, 0x4D34373142353637334648302D4346382020
Memory Module: BANK 1/DIMM0, 2 GB, DDR3, 1067 MHz, 0x80CE, 0x4D34373142353637334648302D4346382020
AirPort: spairport_wireless_card_type_airport_extreme (0x14E4, 0x93), Broadcom BCM43xx 1.0 (5.106.98.102.30)
Bluetooth: Version 6.0.7f22, 3 services, 27 devices, 1 incoming serial ports
Network Service: Wi-Fi, AirPort, en1
Serial ATA Device: Hitachi HTS545032B9SA02, 320.07 GB
Serial ATA Device: OPTIARC DVD RW AD-5680H
USB Device: USB 2.0 Bus
USB Device: USB 2.0 Bus
USB Device: Keyboard Hub
USB Device: Go To Final Lap
USB Device: Apple Keyboard
USB Device: JMS579
USB Device: USB Bus
USB Device: USB Bus
USB Device: BRCM2070 Hub
USB Device: Bluetooth USB Host Controller
USB Device: IR Receiver
Thunderbolt Bus: 
```

### Additional Notes

I tried doing `sysctl -a > /Users/haravikk/Desktop/sysctl.txt` but the panic seems to prevent anything from being written out in time (file is empty), so I'm afraid I can't give you a ballpark for which entry it might be.

My guess would be that it's something to do with one the options that presents live data from the kernel, perhaps one of the new ones?

**Comment by lundman** [comment]
Thanks for reporting it, was in the new cpuid code. I have posted rc4.

**Comment by Haravikk** [comment]
That seems to have sorted it! Here's the CPUID logging for reference on my 2009 Mac Mini:

```
2024-03-20 10:45:21.579 Df kernel[0:62d6] (zfs) <> ZFS: Loading module ...
2024-03-20 10:45:21.579 Df kernel[0:62d6] (zfs) <> SPL: loading
2024-03-20 10:45:21.640 Df kernel[0:62d6] (zfs) <> CPUID: sse sse2 sse3 ssse3 sse4.1
2024-03-20 10:45:21.640 Df kernel[0:62d6] (zfs) <> CPUID:
2024-03-20 10:45:23.145 Df kernel[0:62d6] (zfs) <> ZFS filesystem version: 5
2024-03-20 10:45:23.161 Df kernel[0:62d6] (zfs) <> ZFS: Loaded module v2.2.3-rc4, ZFS pool version 5000, ZFS filesystem version 5
```

This all seems consistent with what I expect from this system, and I'm only seeing the implementations I expect as well:

```
sysctl -a | grep '_impl:'
kstat.zfs.darwin.tunable.zfs_vdev_raidz_impl: cycle [fastest] original scalar sse2 ssse3 
kstat.zfs.darwin.tunable.icp_aes_impl: cycle [fastest] generic x86_64 
kstat.zfs.darwin.tunable.icp_gcm_impl: cycle [fastest] generic 
kstat.zfs.darwin.tunable.zfs.fletcher_4_impl: [fastest] scalar superscalar superscalar4 sse2 ssse3 
kstat.zfs.darwin.tunable.zfs.blake3_impl: cycle [fastest] generic sse2 sse41 
kstat.zfs.darwin.tunable.zfs.sha256_impl: cycle [fastest] generic x64 ssse3 
kstat.zfs.darwin.tunable.zfs.sha512_impl: cycle [fastest] generic x64 
```

I won't mark this closed just yet as I'd also like to double check my other systems, specifically my 2018 Mac Mini, but that's my main working machine so I often have things running I can't just interrupt, so it'll be a few days again.

I also want to try it against [the issues I've had with performance on that system](https://openzfsonosx.org/forum/viewtopic.php?p=12758#p12758) when using newer releases. On which note, if rc4 doesn't help with that, would you prefer me to move that problem to a new issue here? Another user (armdn) has confirmed I'm not alone in experiencing poor performance on releases after v2.1.0.

**Comment by Haravikk** [comment]
I've confirmed that this issue is fixed under v2.2.3rc4 for my 2018 Mac Mini as well, and I'm seeing all of the expected CPUID features/implementations:

```
$ log show --last 10m --source --predicate 'sender == "zfs"' --style compact
2024-03-24 11:46:51.963 Df kernel[0:3e71] (zfs) <zfs`org_openzfsonosx_zfs_zvol::start(IOService*)> ZFS: Loading module ...
2024-03-24 11:46:51.963 Df kernel[0:3e71] (zfs) <zfs`spl_start> SPL: loading
2024-03-24 11:46:51.978 Df kernel[0:3e71] (zfs) <zfs`spl_processor_init> CPUID: osxsave sse sse2 sse3 ssse3 sse4.1 sse4.2
2024-03-24 11:46:51.978 Df kernel[0:3e71] (zfs) <zfs`spl_processor_init> CPUID: avx avx2 aes pclmulqdq movbe
2024-03-24 11:46:51.978 Df kernel[0:3e71] (zfs) <zfs`org_openzfsonosx_zfs_zvol::start(IOService*)> ZFS: hostid set to ac938a5b from UUID '98A28080-37FC-5F46-B29E-EBFD29E61994'
2024-03-24 11:46:52.893 Df kernel[0:3e71] (zfs) <zfs`zfs_init> ZFS filesystem version: 5
2024-03-24 11:46:52.908 Df kernel[0:3e71] (zfs) <zfs`org_openzfsonosx_zfs_zvol::start(IOService*)> ZFS: Loaded module v2.2.3-rc4, ZFS pool version 5000, ZFS filesystem version 5
```

```
$sysctl -a | grep '_impl:'
kstat.zfs.darwin.tunable.zfs_vdev_raidz_impl: cycle [fastest] original scalar sse2 ssse3 avx2 
kstat.zfs.darwin.tunable.icp_aes_impl: cycle [fastest] generic x86_64 aesni 
kstat.zfs.darwin.tunable.icp_gcm_impl: cycle [fastest] avx generic pclmulqdq 
kstat.zfs.darwin.tunable.zfs.fletcher_4_impl: [fastest] scalar superscalar superscalar4 sse2 ssse3 avx2 
kstat.zfs.darwin.tunable.zfs.blake3_impl: cycle [fastest] generic sse2 sse41 avx2 
kstat.zfs.darwin.tunable.zfs.sha256_impl: cycle [fastest] generic x64 ssse3 avx avx2 
kstat.zfs.darwin.tunable.zfs.sha512_impl: cycle [fastest] generic x64 avx avx2
```

So I'm going to close this as it seems to be fully resolved!
